File: Data Structures & Algorithms/house-robber/submission-1.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        
        int len = nums.size();

        int twoHousesBehind = 0;
        int oneHouseBehind = nums[0];
        int currMaxVal = nums[0];

        for(int i=2;i<=len;i++)
        {
            currMaxVal = max(twoHousesBehind + nums[i-1], oneHouseBehind);
            twoHousesBehind = oneHouseBehind;
            oneHouseBehind = currMaxVal;
        }

        return currMaxVal;
    }
};

File: Data Structures & Algorithms/maximum-product-subarray/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    maxProduct(nums) {

        let prefix = 1, suffix = 1, maxResult = -Infinity;
        const len = nums.length;

        for(let i=0;i<len;i++)
        {
            if(prefix === 0 || prefix == -0) prefix = 1;
            if(suffix === 0 || suffix == -0) suffix = 1;

            prefix = prefix * nums[i];
            suffix = suffix * nums[len-i-1];

            maxResult = Math.max(maxResult, Math.max(prefix, suffix));
        }

        return maxResult == -0 ? 0 : maxResult;
    }
}
